numpy/core/src/multiarray/scalartypes.c.src

```diff
@@ -2627,14 +2615,6 @@ static PyObject *
     Py_DECREF(arr);
 
 finish:
-    /*
-     * In OBJECT case, robj is no longer a
-     * PyArrayScalar at this point but the
-     * remaining code assumes it is
-     */
-#if defined(_@TYPE@_IS_OBJECT)
-    return robj;
-#else
     /* Normal return */
     if ((robj == NULL) || (Py_TYPE(robj) == type)) {
         return robj;
```

Comment:
Adding an assert after the below comment might be nice, since only float, string and unicode can reach it I guess? (but no priority)

Reply:
A good idea, but something I want to leave till after this patch.

If you want to take a stab at cleaning this up further in a follow-up, please do!